File: OrcScala/src/orc/lib/state/Counter.java
//
// Counter.java -- Java class Counter
// Project OrcScala
//
// Copyright (c) 2018 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state;

import java.math.BigDecimal;
import java.util.LinkedList;

import orc.values.sites.compatibility.CallContext;
import orc.MaterializedCallContext;
import orc.error.runtime.ArityMismatchException;
import orc.error.runtime.TokenException;
import orc.lib.state.types.CounterType;
import orc.run.distrib.AbstractLocation;
import orc.run.distrib.ClusterLocations;
import orc.run.distrib.DOrcPlacementPolicy;
import orc.types.Type;
import orc.values.sites.TypedSite;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.DotSite;
import orc.values.sites.compatibility.EvalSite;
import orc.values.sites.compatibility.PartialSite;
import orc.values.sites.compatibility.SiteAdaptor;

/**
 * Factory for counters.
 *
 * @author quark
 */
public class Counter extends EvalSite implements TypedSite {
    @Override
    public Object evaluate(final Args args) throws TokenException {
        final int init = args.size() == 0 ? 0 : args.intArg(0);

        if (args.size() > 1) {
            // technically of arity 0 or 1
            throw new ArityMismatchException(1, args.size());
        }

        return new CounterInstance(init);
    }

    @Override
    public Type orcType() {
        return CounterType.getBuilder();
    }

    protected class CounterInstance extends DotSite implements DOrcPlacementPolicy {
        // TODO: Reimplement this without the lock. It will probably scale much better with AtomicInteger
        protected int count;
        protected final LinkedList<MaterializedCallContext> waiters = new LinkedList<MaterializedCallContext>();

        public CounterInstance(final int init) {
            super();
            count = init;
        }

        @Override
        protected void addMembers() {
            addMember("inc", new EvalSite() {
                @Override
                public Object evaluate(final Args args) throws TokenException {
                    synchronized (Counter.this) {
                        ++count;
                    }
                    return signal();
                }

                @Override
                public boolean nonBlocking() {
                    return true;
                }
            });
            addMember("dec", new PartialSite() {
                @Override
                public Object evaluate(final Args args) throws TokenException {
                    synchronized (Counter.this) {
                        if (count > 0) {
                            --count;
                            if (count == 0) {
                                LinkedList<CallContext> oldWaiters = (LinkedList<CallContext>) waiters.clone();
                                waiters.clear();
                                for (final CallContext waiter : oldWaiters) {
                                    waiter.publish(signal());
                                }
                            }
                            return signal();
                        } else {
                            return null;
                        }
                    }
                }

                @Override
                public boolean nonBlocking() {
                    return true;
                }
            });
            addMember("onZero", new SiteAdaptor() {
                @Override
                public void callSite(final Args args, final CallContext caller) throws TokenException {
                    synchronized (Counter.this) {
                        if (count == 0) {
                            caller.publish(signal());
                        } else {
                            caller.setQuiescent();
                            waiters.add(caller.materialize());
                        }
                    }
                }
            });
            addMember("value", new EvalSite() {
                @Override
                public Object evaluate(final Args args) throws TokenException {
                    return BigDecimal.valueOf(count);
                }

                @Override
                public boolean nonBlocking() {
                    return true;
                }
            });
        }

        @Override
        public <L extends AbstractLocation> scala.collection.immutable.Set<L> permittedLocations(final ClusterLocations<L> locations) {
            return locations.hereSet();
        }

    }

    @Override
    public boolean nonBlocking() {
        return true;
    }

    @Override
    public int minPublications() {
        return 0;
    }

    @Override
    public int maxPublications() {
        return 1;
    }

    @Override
    public boolean effectFree() {
        return true;
    }
}

File: OrcScala/src/orc/lib/state/Counter.scala
//
// Counter.java -- Java class Counter
// Project OrcScala
//
// Copyright (c) 2017 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state

import java.util.LinkedList

import scala.collection.JavaConverters.asScalaBufferConverter

import orc.{MaterializedCallContext, VirtualCallContext}
import orc.lib.state.types.CounterType
import orc.run.distrib.{AbstractLocation, ClusterLocations, DOrcPlacementPolicy}
import orc.types.Type
import orc.values.{FastObject, NumericsConfig, Signal}
import orc.values.sites.{
  TalkativeSite,
  NonBlockingSite,
  Site0Simple,
  SiteMetadata,
  TotalSite0Simple,
  TotalSite1Simple,
  TypedSite
}

/**
  * Factory for counters.
  *
  * @author quark
  */
object Counter
    extends TotalSite1Simple[Number]
    with TypedSite
    with NonBlockingSite with TalkativeSite {
  override def eval(init: Number): AnyRef = {
    new Counter.Instance(init.intValue)
  }

  override def orcType(): Type = CounterType.getBuilder

  private val members = FastObject.members("inc", "dec", "onZero", "value")

  class Instance(protected var count: Int)
      extends FastObject(members)
      with DOrcPlacementPolicy {

    protected val waiters: LinkedList[MaterializedCallContext] =
      new LinkedList[MaterializedCallContext]()

    val values = Array(
      // "inc",
      new TotalSite0Simple with NonBlockingSite {
        override def eval(): AnyRef = {
          Counter.this.synchronized { count += 1 }
          Signal
        }
      },
      // "dec",
      new Site0Simple with NonBlockingSite {
        override def eval(ctx: VirtualCallContext) =
          Counter.this.synchronized {
            if (count > 0) {
              val r1 = ctx.publish(Signal)
              count -= 1
              if (count == 0) {
                val oldWaiters = waiters.asScala.toArray
                waiters.clear()
                oldWaiters.foldLeft(r1)(_.publish(_, Signal))
              } else r1
            } else {
              ctx.empty
            }
          }
      },
      // "onZero",
      new Site0Simple with SiteMetadata {
        override def eval(caller: VirtualCallContext) =
          Counter.this.synchronized {
            if (count == 0) {
              caller.publish(Signal)
            } else {
              val mat = caller.materialize()
              mat.setQuiescent()
              waiters.add(mat)
              caller.empty
            }
          }
      },
      // "value",
      new TotalSite0Simple with NonBlockingSite {
        override def eval(): AnyRef = NumericsConfig.toOrcIntegral(count)
      },
    )

    override def permittedLocations[L <: AbstractLocation](
        locations: ClusterLocations[L]): scala.collection.immutable.Set[L] =
      locations.hereSet
  }
}
